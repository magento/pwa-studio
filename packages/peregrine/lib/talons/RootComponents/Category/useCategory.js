import { useEffect, useMemo, useRef } from 'react';
import { useLocation } from 'react-router-dom';
import { useLazyQuery, useQuery } from '@apollo/client';

import mergeOperations from '../../../util/shallowMerge';
import { useAppContext } from '../../../context/app';
import { usePagination } from '../../../hooks/usePagination';
import { useScrollTopOnChange } from '../../../hooks/useScrollTopOnChange';
import { useSort } from '../../../hooks/useSort';
import {
    getFiltersFromSearch,
    getFilterInput
} from '../../../talons/FilterModal/helpers';

import DEFAULT_OPERATIONS from './category.gql';

/**
 * A [React Hook]{@link https://reactjs.org/docs/hooks-intro.html} that
 * controls the logic for the Category Root Component.
 *
 * @kind function
 *
 * @param {object}      props
 * @param {number}      props.id - Category Id.
 * @param {GraphQLAST}  props.operations.getCategoryQuery - Fetches category using a server query
 * @param {GraphQLAST}  props.operations.getFilterInputsQuery - Fetches "allowed" filters using a server query
 * @param {GraphQLAST}  props.queries.getStoreConfig - Fetches store configuration using a server query
 *
 * @returns {object}    result
 * @returns {object}    result.error - Indicates a network error occurred.
 * @returns {object}    result.categoryData - Category data.
 * @returns {bool}      result.isLoading - Category data loading.
 * @returns {string}    result.metaDescription - Category meta description.
 * @returns {object}    result.pageControl - Category pagination state.
 * @returns {array}     result.sortProps - Category sorting parameters.
 * @returns {number}    result.pageSize - Category total pages.
 */
export const useCategory = props => {
    const {
        id,
        queries: { getPageSize }
    } = props;

    const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);
    const { getCategoryQuery, getFilterInputsQuery } = operations;

    const { data: pageSizeData } = useQuery(getPageSize, {
        fetchPolicy: 'cache-and-network',
        nextFetchPolicy: 'cache-first'
    });
    const pageSize = pageSizeData && pageSizeData.storeConfig.grid_per_page;

    const [paginationValues, paginationApi] = usePagination();
    const { currentPage, totalPages } = paginationValues;
    const { setCurrentPage, setTotalPages } = paginationApi;

    const sortProps = useSort();
    const [currentSort] = sortProps;

    // Keep track of the sort criteria so we can tell when they change.
    const previousSort = useRef(currentSort);

    const pageControl = {
        currentPage,
        setPage: setCurrentPage,
        totalPages
    };

    const [
        ,
        {
            actions: { setPageLoading }
        }
    ] = useAppContext();

    const [runQuery, queryResponse] = useLazyQuery(getCategoryQuery, {
        fetchPolicy: 'cache-and-network',
        nextFetchPolicy: 'cache-first'
    });
    const {
        called: categoryCalled,
        loading: categoryLoading,
        error,
        data
    } = queryResponse;
    const { search } = useLocation();

    const isBackgroundLoading = !!data && categoryLoading;

    // Update the page indicator if the GraphQL query is in flight.
    useEffect(() => {
        setPageLoading(isBackgroundLoading);
    }, [isBackgroundLoading, setPageLoading]);

    // Keep track of the search terms so we can tell when they change.
    const previousSearch = useRef(search);

    // Get "allowed" filters by intersection of schema and aggregations
    const {
        called: introspectionCalled,
        data: introspectionData,
        loading: introspectionLoading
    } = useQuery(getFilterInputsQuery, {
        fetchPolicy: 'cache-and-network',
        nextFetchPolicy: 'cache-first'
    });

    // Create a type map we can reference later to ensure we pass valid args
    // to the graphql query.
    // For example: { category_id: 'FilterEqualTypeInput', price: 'FilterRangeTypeInput' }
    const filterTypeMap = useMemo(() => {
        const typeMap = new Map();
        if (introspectionData) {
            introspectionData.__type.inputFields.forEach(({ name, type }) => {
                typeMap.set(name, type.name);
            });
        }
        return typeMap;
    }, [introspectionData]);

    // Run the category query immediately and whenever its variable values change.
    useEffect(() => {
        // Wait until we have the type map to fetch product data.
        if (!filterTypeMap.size || !pageSize) {
            return;
        }

        const filters = getFiltersFromSearch(search);

        // Construct the filter arg object.
        const newFilters = {};
        filters.forEach((values, key) => {
            newFilters[key] = getFilterInput(values, filterTypeMap.get(key));
        });

        // Use the category id for the current category page regardless of the
        // applied filters. Follow-up in PWA-404.
        newFilters['category_id'] = { eq: String(id) };

        runQuery({
            variables: {
                currentPage: Number(currentPage),
                id: String(id),
                filters: newFilters,
                pageSize: Number(pageSize),
                sort: { [currentSort.sortAttribute]: currentSort.sortDirection }
            }
        });
    }, [
        currentPage,
        currentSort,
        filterTypeMap,
        id,
        pageSize,
        runQuery,
        search
    ]);

    const totalPagesFromData = data
        ? data.products.page_info.total_pages
        : null;

    useEffect(() => {
        setTotalPages(totalPagesFromData);
        return () => {
            setTotalPages(null);
        };
    }, [setTotalPages, totalPagesFromData]);

    // If we get an error after loading we should try to reset to page 1.
    // If we continue to have errors after that, render an error message.
    useEffect(() => {
        if (error && !categoryLoading && !data && currentPage !== 1) {
            setCurrentPage(1);
        }
    }, [currentPage, error, categoryLoading, setCurrentPage, data]);

    // Reset the current page back to one (1) when the search string, filters
    // or sort criteria change.
    useEffect(() => {
        // We don't want to compare page value.
        const prevSearch = new URLSearchParams(previousSearch.current);
        const nextSearch = new URLSearchParams(search);
        prevSearch.delete('page');
        nextSearch.delete('page');

        if (
            prevSearch.toString() !== nextSearch.toString() ||
            previousSort.current.sortAttribute.toString() !==
                currentSort.sortAttribute.toString() ||
            previousSort.current.sortDirection.toString() !==
                currentSort.sortDirection.toString()
        ) {
            // The search term changed.
            setCurrentPage(1, true);
            // And update the ref.
            previousSearch.current = search;
            previousSort.current = currentSort;
        }
    }, [currentSort, previousSearch, search, setCurrentPage]);

    const categoryData = categoryLoading && !data ? null : data;
    const metaDescription =
        data &&
        data.categories.items[0] &&
        data.categories.items[0].meta_description
            ? data.categories.items[0].meta_description
            : '';

    // When only categoryLoading is involved, noProductsFound component flashes for a moment
    const loading =
        (introspectionCalled && !categoryCalled) ||
        (categoryLoading && !data) ||
        introspectionLoading;

    useScrollTopOnChange(currentPage);

    return {
        error,
        categoryData,
        loading,
        metaDescription,
        pageControl,
        sortProps,
        pageSize
    };
};
