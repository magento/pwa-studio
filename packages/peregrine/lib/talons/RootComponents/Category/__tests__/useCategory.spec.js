import React from 'react';
import { act } from 'react-test-renderer';
import createTestInstance from '@magento/peregrine/lib/util/createTestInstance';
import { useCategory } from '../useCategory';
import { useQuery, useLazyQuery } from '@apollo/client';
import { useLocation } from 'react-router-dom';

jest.mock('react-router-dom', () => ({
    useHistory: jest.fn(() => ({ push: jest.fn() })),
    useLocation: jest.fn(() => ({ pathname: '', search: '' }))
}));

jest.mock('../../../../hooks/useScrollTopOnChange');

jest.mock('../../../../context/app', () => {
    const state = {};
    const api = {
        actions: { setPageLoading: jest.fn() }
    };
    const useAppContext = jest.fn(() => [state, api]);

    return { useAppContext };
});

const mockUseSort = jest
    .fn()
    .mockReturnValue([
        {
            sortText: 'Best Match',
            sortAttribute: 'relevance',
            sortDirection: 'DESC'
        },
        jest.fn()
    ])
    .mockName('mockUseSort');

const mockSetCurrentPage = jest.fn().mockName('mockSetCurrentPage');

jest.mock('../../../../hooks/usePagination', () => ({
    usePagination: jest.fn(() => [
        {
            currentPage: 3,
            totalPages: 6
        },
        {
            setCurrentPage: jest
                .fn()
                .mockImplementation((page, bool = false) =>
                    mockSetCurrentPage(page, bool)
                ),
            setTotalPages: jest.fn()
        }
    ])
}));

jest.mock('../../../../hooks/useSort', () => ({
    useSort: jest.fn().mockImplementation(() => mockUseSort())
}));

jest.mock('@apollo/client', () => {
    const apolloClient = jest.requireActual('@apollo/client');
    const useQuery = jest.fn(() => ({
        called: false,
        data: null,
        loading: false
    }));
    const useLazyQuery = jest.fn();

    return { ...apolloClient, useLazyQuery, useQuery };
});

const mockProps = {
    id: 7,
    queries: {}
};

const mockPageSizeData = {
    called: true,
    data: {
        __type: {
            inputFields: []
        },
        storeConfig: {
            grid_per_page: 12
        }
    },
    error: null,
    loading: false
};

const mockFilterInputsData = {
    called: true,
    error: null,
    loading: false,
    data: {
        __type: {
            inputFields: [
                {
                    name: 'category_id',
                    type: {
                        name: 'FilterEqualTypeInput'
                    }
                },
                {
                    name: 'price',
                    type: {
                        name: 'FilterRangeTypeInput'
                    }
                }
            ]
        }
    }
};

const mockCategoryData = {
    called: true,
    error: null,
    loading: false,
    data: {
        categories: {
            items: [
                {
                    meta_description: 'Category meta-description'
                }
            ]
        },
        products: {
            page_info: {
                total_pages: 6
            }
        }
    }
};

const mockRunQuery = jest.fn();

const Component = props => {
    const talonProps = useCategory(props);
    return <i talonProps={talonProps} />;
};
describe('useCategory tests', () => {
    test('returns the correct shape', () => {
        useQuery.mockReturnValue(mockPageSizeData);
        useLazyQuery.mockReturnValue([mockRunQuery, mockCategoryData]);

        const tree = createTestInstance(<Component {...mockProps} />);
        const { root } = tree;
        const { talonProps } = root.findByType('i').props;

        expect(talonProps).toMatchSnapshot();
    });

    test('runs the category query', () => {
        useLazyQuery.mockReturnValue([mockRunQuery, mockCategoryData]);
        useQuery
            .mockReturnValueOnce(mockPageSizeData)
            .mockReturnValueOnce(mockFilterInputsData);

        useLocation.mockReturnValue({
            pathname: '',
            search: 'page=1&price%5Bfilter%5D=0-100%2C0_100'
        });

        createTestInstance(<Component {...mockProps} />);

        expect(mockRunQuery).toHaveBeenCalledTimes(1);
        expect(mockRunQuery.mock.calls[0][0]).toMatchSnapshot();
    });

    test('resets the current page on error', () => {
        useLazyQuery.mockReturnValue([
            mockRunQuery,
            {
                loading: false,
                error: {
                    message: 'An error ocurred!'
                },
                data: null
            }
        ]);
        useQuery
            .mockReturnValueOnce(mockPageSizeData)
            .mockReturnValueOnce(mockFilterInputsData);

        createTestInstance(<Component {...mockProps} />);

        expect(mockSetCurrentPage).toHaveBeenCalledTimes(1);
        expect(mockSetCurrentPage).toHaveBeenCalledWith(1, false);
    });

    test('handles no filter type data available', () => {
        useLazyQuery.mockReturnValue([mockRunQuery, mockCategoryData]);
        useQuery.mockReturnValueOnce(mockPageSizeData).mockReturnValueOnce({
            error: null,
            loading: true,
            data: null
        });
        createTestInstance(<Component {...mockProps} />);

        expect(mockRunQuery).toHaveBeenCalledTimes(0);
    });

    test('category query loading state', () => {
        useLazyQuery.mockReturnValue([
            mockRunQuery,
            {
                loading: true,
                error: null,
                data: null
            }
        ]);
        useQuery
            .mockReturnValueOnce(mockPageSizeData)
            .mockReturnValueOnce(mockFilterInputsData);

        const tree = createTestInstance(<Component {...mockProps} />);
        const { root } = tree;
        const { talonProps } = root.findByType('i').props;

        const { loading, categoryData } = talonProps;

        expect(loading).toBeTruthy();
        expect(categoryData).toBeNull();
    });

    test('loading state when only categoryLoading is involved', () => {
        useLazyQuery.mockReturnValue([
            mockRunQuery,
            {
                called: true,
                loading: true,
                error: null,
                data: null
            }
        ]);
        useQuery.mockReturnValueOnce(mockPageSizeData).mockReturnValueOnce({
            called: false,
            loading: false,
            error: null,
            data: null
        });

        const tree = createTestInstance(<Component {...mockProps} />);
        const { root } = tree;
        const { talonProps } = root.findByType('i').props;

        const { loading } = talonProps;

        expect(loading).toBeTruthy();
    });

    test('sets current page to 1 if error, !loading, !data, and currentPage != 1', () => {
        useLazyQuery.mockReturnValue([
            mockRunQuery,
            {
                ...mockCategoryData,
                loading: false,
                data: null,
                error: true
            }
        ]);

        createTestInstance(<Component {...mockProps} />);

        expect(mockSetCurrentPage).toHaveBeenCalledWith(1, false);
    });

    const testCases = [
        [
            'sortText does not reset',
            {
                sortText: 'Changed',
                sortAttribute: 'relevance',
                sortDirection: 'DESC'
            },
            0
        ],
        [
            'sortAttribute resets',
            {
                sortText: 'Best Match',
                sortAttribute: 'Changed',
                sortDirection: 'DESC'
            },
            1
        ],
        [
            'sortDirection resets',
            {
                sortText: 'Best Match',
                sortAttribute: 'relevance',
                sortDirection: 'Changed'
            },
            1
        ]
    ];

    test.each(testCases)(
        'Changing %s current page to 1.',
        (description, sortParams, expected) => {
            useQuery.mockReturnValue(mockPageSizeData);
            useLazyQuery.mockReturnValue([mockRunQuery, mockCategoryData]);

            const tree = createTestInstance(<Component {...mockProps} />);

            mockUseSort.mockReturnValueOnce([sortParams, jest.fn()]);
            act(() => {
                tree.update(<Component {...mockProps} />);
            });

            expect(mockSetCurrentPage).toHaveBeenCalledTimes(expected);
        }
    );

    test('preserve history when search term changes', () => {
        useQuery.mockReturnValue(mockPageSizeData);
        useLazyQuery.mockReturnValue([mockRunQuery, mockCategoryData]);

        const tree = createTestInstance(<Component {...mockProps} />);

        mockUseSort.mockReturnValueOnce([
            {
                sortText: 'Best Match',
                sortAttribute: 'relevance',
                sortDirection: 'Changed'
            },
            jest.fn()
        ]);
        expect(mockSetCurrentPage).not.toHaveBeenCalledWith(1, true);
        act(() => {
            tree.update(<Component {...mockProps} />);
        });
        expect(mockSetCurrentPage).toHaveBeenCalledWith(1, true);
    });
});
