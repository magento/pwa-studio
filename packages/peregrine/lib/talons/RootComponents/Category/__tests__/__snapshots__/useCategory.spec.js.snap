// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns the correct shape 1`] = `
Object {
  "categoryData": Object {
    "category": Object {
      "meta_description": "Category meta-description",
    },
    "products": Object {
      "page_info": Object {
        "total_pages": 6,
      },
    },
  },
  "error": null,
  "loading": false,
  "metaDescription": "Category meta-description",
  "pageControl": Object {
    "currentPage": 3,
    "setPage": [MockFunction],
    "totalPages": 6,
  },
  "pageSize": 12,
  "sortProps": Array [
    Object {
      "sortAttribute": "relevance",
      "sortDirection": "DESC",
      "sortText": "Best Match",
    },
    [MockFunction],
  ],
}
`;

exports[`runs the category query 1`] = `
Object {
  "variables": Object {
    "currentPage": 3,
    "filters": Object {
      "category_id": Object {
        "eq": "7",
      },
      "price": Object {
        "from": "0",
        "to": "100",
      },
    },
    "id": 7,
    "pageSize": 12,
    "sort": Object {
      "relevance": "DESC",
    },
  },
}
`;

exports[`useCategory tests returns the correct shape 1`] = `
Object {
  "categoryData": Object {
    "categories": Object {
      "items": Array [
        Object {
          "meta_description": "Category meta-description",
        },
      ],
    },
    "products": Object {
      "page_info": Object {
        "total_pages": 6,
      },
    },
  },
  "error": null,
  "loading": false,
  "metaDescription": "Category meta-description",
  "pageControl": Object {
    "currentPage": 3,
    "setPage": [MockFunction],
    "totalPages": 6,
  },
  "pageSize": 12,
  "sortProps": Array [
    Object {
      "sortAttribute": "relevance",
      "sortDirection": "DESC",
      "sortText": "Best Match",
    },
    [MockFunction],
  ],
}
`;

exports[`useCategory tests runs the category query 1`] = `
Object {
  "variables": Object {
    "currentPage": 3,
    "filters": Object {
      "category_id": Object {
        "eq": "7",
      },
      "price": Object {
        "from": "0",
        "to": "100",
      },
    },
    "id": "7",
    "pageSize": 12,
    "sort": Object {
      "relevance": "DESC",
    },
  },
}
`;
